packages/editor/src/components/post-author/index.js

```diff
@@ -16,34 +23,146 @@ export class PostAuthor extends Component {
+		this.suggestAuthor = this.suggestAuthor.bind( this );
+		this.getCurrentAuthor = this.getCurrentAuthor.bind( this );
+		this.resolveResults = this.resolveResults.bind( this );
+		this.requestResults = debounce( ( query, populateResults ) => {
+			const payload = '?search=' + encodeURIComponent( query );
+			apiFetch( { path: `/wp/v2/users${ payload }` } ).then( ( results ) => {
+				populateResults( this.resolveResults( results ) );
+				this.searchCache[ query ] = results;
+			} );
+		}, 300 );
+		this.searchCache = [];
+		this.state = {
+			postAuthor: false,
+		};
+	}
+
+	getInitialPostAuthor() {
+		// Get the initial author from props if available.
+		return this.props.postAuthor;
+	}
+
+	componentDidMount() {
+		const { postAuthorId, authors } = this.props;
+
+		// Load the initial post author.
+		const postAuthor = this.getInitialPostAuthor();
+		this.setState( { postAuthor } );
+
+		const authorInAuthors = authors.find( ( singleAuthor ) => {
+			return singleAuthor.id === postAuthorId;
+		} );
+
+		// Set or fetch the current author.
+		if ( authorInAuthors ) {
+			this.setState( { postAuthor: authorInAuthors } );
+		} else {
+			this.getCurrentAuthor( postAuthorId );
+		}
+	}
+
+	suggestAuthor( query, populateResults ) {
+		if ( query.length < 2 ) {
+			return;
+		}
+
+		if ( this.searchCache[ query ] ) {
+			populateResults( this.resolveResults( this.searchCache[ query ] ) );
+			return;
+		}
+
+		this.requestResults( query, populateResults );
 	}
 
-	setAuthorId( event ) {
+	resolveResults( results ) {
+		this.authors = results;
+		return results.map( ( author ) => ( author.name ) );
+	}
+
+	getCurrentAuthor( authorId ) {
+		if ( ! authorId ) {
+			return;
+		}
+		apiFetch( { path: `/wp/v2/users/${ encodeURIComponent( authorId ) }` } ).then( ( results ) => {
+			this.setState( { postAuthor: results } );
+		} );
+	}
+
+	setAuthorId( selection ) {
+		if ( ! selection ) {
+			return;
+		}
 		const { onUpdateAuthor } = this.props;
-		const { value } = event.target;
-		onUpdateAuthor( Number( value ) );
+		if ( typeof selection === 'string' ) {
+			// Author name from the autocompleter.
+			const author = this.authors.find( ( singleAuthor ) => {
+				return singleAuthor.name === selection;
+			} );
+			onUpdateAuthor( Number( author.id ) );
+		} else {
+			// Author ID from the select.
+			onUpdateAuthor( Number( selection.target.value ) );
+		}
 	}
 
 	render() {
-		const { postAuthor, instanceId, authors } = this.props;
-		const selectId = 'post-author-selector-' + instanceId;
-
-		// Disable reason: A select with an onchange throws a warning
+		const { postAuthor } = this.state;
+		const { postAuthorId, instanceId, authors } = this.props;
+		const selectId = `post-author-selector-${ instanceId }`;
+		let selector;
+		if ( ! postAuthor ) {
+			return null;
+		}
 
-		/* eslint-disable jsx-a11y/no-onchange */
-		return (
-			<PostAuthorCheck>
-				<label htmlFor={ selectId }>{ __( 'Author' ) }</label>
+		if ( authors.length < 50 ) {
```

Comment:
Thanks @swissspidy . That leads to a recommendation, but it isn't very clear on the specific reasons around why it would be a good idea to change the behavior depending on the number of options. The specific phrasing reads a bit confused to me as well, being ambiguous on whether or not "you can use `enhanceSelectElement` to enhance it" is meant to apply only to small lists, large lists, or both. It's also not clear to me it's an accessibility point at all, vs. a mere responsible-bandwidth recommendation where more than a few hundred select options would be unreasonably intensive/large to generate server-side.

Reply:
> why it would be a good idea to change the behavior depending on the number of options

It's a good idea because **with a small number of options, a regular HTML select element is an excellent choice**: its concise, accessible, and a fundamental HTML component that all browsers and accessibility aids such as screen readers handle well. 

**Beyond a certain size, however, dropdowns become a poor choice**: they are unwieldy, all elements must be loaded ahead of time before the dropdown can display resulting in a slow load time especially on low bandwidth connections, searching by typing is difficult (pausing often breaks the search), very large dropdowns crash the browser on less powerful devices, searching large dropdowns on mobile is nearly impossible, etc etc.

> It's also not clear to me it's an accessibility point at all, vs. a mere responsible-bandwidth recommendation

Right, this is not an accessibility issue directly. Rather it is a bandwidth, memory, best practice, usability, scalability etc issue. Regarding accessibility: the technical solution I have chosen to address the many issues with a large dropdown **is accessible**.

Reply:
Thanks for the explanation. As long as we're acknowledging the following, I think it may be an agreeable compromise:

- We're stating that `AccessibleAutocomplete` is inferior than a `select` dropdown at rendering small lists of items
- We're accepting an increased maintenance cost in maintaining two separate implementations, and a higher risk of bugs in mind of the fact that a distinct experience is presented on the basis of the number of users on a site (for example, I will rarely encounter AccessibleAutocomplete because my site has fewer than 50 users)

Reply:
Right.

> We're stating that AccessibleAutocomplete is inferior than a select dropdown at rendering small lists of items

AccessibleAutocomplete explicitly recommends this type of **progressive enhancement**: native elements work very well for smaller numbers of elements.

> We're accepting an increased maintenance cost in maintaining two separate implementations

Yes, we are accepting maintaining the handling of both states (the implementations themselves are well maintained elsewhere).

Reply:
I'd totally second this. The autocomplete "filtering" feature doesn't make sense when the amount of options is limited. As a user, I really don't need to filter a list if I can see all the available (few) options. It does help a lot when there are lots of options.